src/auth/src/credentials/external_account.rs

```diff
@@ -349,6 +350,159 @@ impl Builder {
+///     .with_token_url("https://sts.googleapis.com/v1beta/token".to_string())
+///     .with_quota_project_id("my-quota-project")
+///     .with_scopes(vec!["https://www.googleapis.com/auth/devstorage.read_only".to_string()])
+///     .build()
+///     .unwrap();
+/// # });
+/// ```
+/// [SubjectTokenProvider]: crate::credentials::subject_token::SubjectTokenProvider
+pub struct ProgrammaticBuilder {
+    quota_project_id: Option<String>,
+    subject_token_provider: Arc<dyn dynamic::SubjectTokenProvider>,
+    config: ExternalAccountConfigBuilder,
+}
+
+impl ProgrammaticBuilder {
+    /// Creates a new builder that uses the provided [`SubjectTokenProvider`] to
+    /// fetch the third-party subject token.
+    pub fn new(subject_token_provider: Arc<dyn dynamic::SubjectTokenProvider>) -> Self {
+        Self {
+            subject_token_provider,
+            quota_project_id: None,
+            config: ExternalAccountConfigBuilder::default(),
+        }
+    }
+
+    /// Sets the optional [quota project] for this credentials.
+    ///
+    /// In some services, you can use a service account in
+    /// one project for authentication and authorization, and charge
+    /// the usage to a different project. This requires that the
+    /// service account has `serviceusage.services.use` permissions on the quota project.
+    ///
+    /// [quota project]: https://cloud.google.com/docs/quotas/quota-project
+    pub fn with_quota_project_id<S: Into<String>>(mut self, quota_project_id: S) -> Self {
+        self.quota_project_id = Some(quota_project_id.into());
+        self
+    }
+
+    /// Sets the [scopes] for these credentials.
+    /// By default `https://www.googleapis.com/auth/cloud-platform` scope is used.
+    ///
+    /// [scopes]: https://developers.google.com/identity/protocols/oauth2/scopes
+    pub fn with_scopes<I, S>(mut self, scopes: I) -> Self
+    where
+        I: IntoIterator<Item = S>,
+        S: Into<String>,
+    {
+        self.config.scopes(
+            scopes
+                .into_iter()
+                .map(|s| s.into())
+                .collect::<Vec<String>>(),
+        );
+        self
+    }
+
+    /// Sets the required audience for the token exchange.
+    ///
+    /// This is the resource name for the workload identity pool and the provider
+    /// identifier in that pool.
+    pub fn with_audience<S: Into<String>>(mut self, audience: S) -> Self {
+        self.config.audience(audience);
+        self
+    }
+
+    /// Sets the required subject token type.
+    ///
+    /// This is the STS subject token type based on the OAuth 2.0 token exchange spec.
+    pub fn with_subject_token_type<S: Into<String>>(mut self, subject_token_type: S) -> Self {
+        self.config.subject_token_type(subject_token_type);
+        self
+    }
+
+    /// Sets the required token URL for the STS token exchange.
+    pub fn with_token_url<S: Into<String>>(mut self, token_url: S) -> Self {
```

Comment:
I believe we can default to `sts.googleapis.com`. We do not have to force user into providing this. @leosiracusa thoughts?